src/v/raft/log_eviction_stm.cc

```diff
@@ -16,23 +16,140 @@
+    /// with exactly one record within it, the point at which to truncate
+    storage::record_batch_builder builder(
+      model::record_batch_type::prefix_truncate, model::offset(0));
+    auto value = serde::to_iobuf(offset);
+    builder.add_raw_kv(iobuf(), std::move(value));
+    auto batch = std::move(builder).build();
+    const auto start = ss::lowres_clock::now();
+    /// After command replication all that can be guaranteed is that the command
+    /// was replicated
+    auto result = co_await replicate_commands(std::move(batch), deadline, as);
+    if (result) {
+        co_return result;
+    }
+    deadline -= ss::lowres_clock::now() - start;
+
+    try {
+        /// The following will return when the logs start_offset has officially
+        /// been incremented at or above the requested truncated offset on this
+        /// node. No guarantees of data removal can be made at this point.
+        //
+        /// Subtract one to account for shift in offset to maintain semantics
+        /// with start offset, see comment in write_snapshot call below
+        co_await _raft->min_visible_offset_monitor().wait(
+          offset - model::offset{1}, deadline, as);
+    } catch (const ss::timed_out_error&) {
+        co_return errc::timeout;
+    } catch (const ss::abort_requested_exception&) {
+        co_return errc::shutting_down;
+    }
+    co_return errc::success;
 }
 
-ss::future<> log_eviction_stm::stop() { return _gate.close(); }
+ss::future<std::error_code> log_eviction_stm::replicate_commands(
+  model::record_batch batch,
+  ss::lowres_clock::time_point deadline,
+  std::optional<std::reference_wrapper<ss::abort_source>> as) {
+    vlog(
+      _logger.info,
+      "Replicating prefix_truncate command current start offset: {}, current "
+      "last offset: {}",
+      _raft->start_offset(),
+      _raft->last_visible_index());
+
+    auto fut = _raft->replicate(
+      _raft->term(),
+      model::make_memory_record_batch_reader(std::move(batch)),
+      raft::replicate_options{raft::consistency_level::quorum_ack});
+
+    /// Execute the replicate command bound by timeout and cancellable via
+    /// abort_source mechanism
+    result<raft::replicate_result> result{{}};
+    try {
+        if (as) {
+            result = co_await ssx::with_timeout_abortable(
+              std::move(fut), deadline, *as);
+        } else {
+            result = co_await ss::with_timeout(deadline, std::move(fut));
+        }
+    } catch (const ss::timed_out_error&) {
+        result = errc::timeout;
+    }
+
+    if (!result) {
+        vlog(
+          _logger.error,
+          "Failed to replicate prefix_truncate command, reason: {}",
+          result.error());
+        co_return result.error();
+    }
+
+    /// Wait until offset has been committed
+    const auto now = ss::lowres_clock::now();
+    if (now >= deadline) {
+        co_return errc::timeout;
+    }
+    const auto timeout = ss::lowres_clock::now() + (deadline - now);
```

Comment:
This looks redundant.  If `now()` is equal to `now) then the "timeout" is just equal to `deadline`.